### PR TITLE
CB-15297 E2E test for DNS lookups

### PR DESCRIPTION
**Pull request by mhalmy**
Jira:
[https://jira.cloudera.com/browse/CB-15297?page=com.atlassian.jira.plugin.system.issuetabpanels%3Aall-tabpanel](url)
DNS lookups should always work, as every FreeIPA instance has nameserver running and we configure all cluster with all FreeIPA as nameserver

### CODE REVIEW

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/testcase/e2e/freeipa/FreeIpaUpgradeTests.java [hunk 1]
```diff
@@ -301,10 +306,37 @@ private void kinit(SdxTestDto sdxTestDto, SdxClient sdxClient) {
...
+        List<InstanceGroupV4Response> instanceGroups = getInstanceGroups(sdxTestDto, sdxClient);
+        for (InstanceGroupV4Response instanceGroup : instanceGroups) {
+            fqdn = instanceGroup.getMetadata().stream().filter(item -> item.getInstanceGroup().equals("idbroker"))
+                    .findFirst().map(item -> item.getDiscoveryFQDN()).orElse(null);
+            privateIP = instanceGroup.getMetadata().stream().filter(item -> item.getInstanceGroup().equals("idbroker"))
+                    .findFirst().map(item -> item.getPrivateIp()).orElse(null);
+            if (fqdn != null) {
+                break;
+            }
+        }
```
Comment (lacikaaa) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/testcase/e2e/freeipa/FreeIpaUpgradeTests.java:321]: I understand this a test but could you refactor this? you should be able to use a single stream to find the idbroker instancemetadata, and throw a `TestFailException` if no suitable found

**Review by lacikaaa — CHANGES_REQUESTED** [review]

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/testcase/e2e/freeipa/FreeIpaUpgradeTests.java (hunk 1 above)
Comment (mhalmy) [reply integration-test/src/main/java/com/sequenceiq/it/cloudbreak/testcase/e2e/freeipa/FreeIpaUpgradeTests.java]: Done.

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/testcase/e2e/freeipa/FreeIpaUpgradeTests.java [hunk 2]
```diff
@@ -301,10 +306,37 @@ private void kinit(SdxTestDto sdxTestDto, SdxClient sdxClient) {
...
+        try {
+            String cmd = CHECK_DNS_LOOKUPS_CMD.replace("##ADDR##", fqdn).replace("##IP##", privateIP);
```
[the remark this answers is not in the document]
Comment (mhalmy) [reply integration-test/src/main/java/com/sequenceiq/it/cloudbreak/testcase/e2e/freeipa/FreeIpaUpgradeTests.java]: Done.

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/testcase/e2e/freeipa/FreeIpaUpgradeTests.java [hunk 3]
```diff
@@ -301,10 +306,37 @@ private void kinit(SdxTestDto sdxTestDto, SdxClient sdxClient) {
...
+            Map<String, Pair<Integer, String>> result = sshJClientActions.executeSshCommand(getInstanceGroups(sdxTestDto, sdxClient),
+                    List.of(HostGroupType.MASTER.getName()), cmd, false);
+            String status = result.entrySet().iterator().next().getValue().getRight();
+            Assertions.assertTrue(status.contains("3"));
```
[the remark this answers is not in the document]
Comment (mhalmy) [reply integration-test/src/main/java/com/sequenceiq/it/cloudbreak/testcase/e2e/freeipa/FreeIpaUpgradeTests.java]: Done.

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/testcase/e2e/freeipa/FreeIpaUpgradeTests.java [hunk 4]
```diff
@@ -301,10 +308,40 @@ private void kinit(SdxTestDto sdxTestDto, SdxClient sdxClient) {
...
+        try {
+            String cmd = String.format(CHECK_DNS_LOOKUPS_CMD, fqdn, privateIP);
+            Map<String, Pair<Integer, String>> result = sshJClientActions.executeSshCommand(getInstanceGroups(sdxTestDto, sdxClient),
+                    List.of(HostGroupType.MASTER.getName()), cmd, false);
+            Iterator<Map.Entry<String, Pair<Integer, String>>> resultIterator = result.entrySet().iterator();
+            while (resultIterator.hasNext()) {
+                String status = resultIterator.next().getValue().getRight();
+                Assertions.assertTrue(status.contains("3"));
+                resultIterator.remove();
+            }
```
Comment (lacikaaa) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/testcase/e2e/freeipa/FreeIpaUpgradeTests.java:332]: this code is still a bit bloated:
```
        InstanceMetaDataV4Response instanceGroupMetadata = getInstanceGroups(sdxTestDto, sdxClient).stream()
                .flatMap(instanceGroup -> instanceGroup.getMetadata().stream())
                .filter(metadata -> metadata.getInstanceGroup().equals("idbroker"))
                .filter(metadata -> StringUtils.isNoneBlank(metadata.getDiscoveryFQDN(), metadata.getPrivateIp()))
                .findFirst().orElseThrow(() -> new TestFailException("FreeIPA upgrade DNS lookups test failed, idbroker instance group was not found."));
        try {
            String cmd = String.format(CHECK_DNS_LOOKUPS_CMD, instanceGroupMetadata.getDiscoveryFQDN(), instanceGroupMetadata.getPrivateIp());
            Map<String, Pair<Integer, String>> results = sshJClientActions.executeSshCommand(getInstanceGroups(sdxTestDto, sdxClient),
                    List.of(HostGroupType.MASTER.getName()), cmd, false);
            results.values().forEach(result -> Assertions.assertTrue(result.getRight().contains("3")));
```
I'm not sure why is it good that we count the IP's on ssh side and then assert on that value. I think the only important thing is if the IP is in the result.
`grep -q $IP` would return with `0` if it's present, and a different value if not. So I would either use this to validate, or bring back the result of the ping to assert if it contains the IP.

**Review by lacikaaa — CHANGES_REQUESTED** [review]

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/testcase/e2e/freeipa/FreeIpaUpgradeTests.java (hunk 4 above)
Comment (mhalmy) [reply integration-test/src/main/java/com/sequenceiq/it/cloudbreak/testcase/e2e/freeipa/FreeIpaUpgradeTests.java]: Done and updated the PR.

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/testcase/e2e/freeipa/FreeIpaUpgradeTests.java [hunk 5]
```diff
@@ -301,10 +308,27 @@ private void kinit(SdxTestDto sdxTestDto, SdxClient sdxClient) {
...
+        InstanceMetaDataV4Response instanceGroupMetadata = getInstanceGroups(sdxTestDto, sdxClient).stream()
+                .flatMap(instanceGroup -> instanceGroup.getMetadata().stream())
+                .filter(metadata -> metadata.getInstanceGroup().equals("idbroker"))
+                .filter(metadata -> StringUtils.isNoneBlank(metadata.getDiscoveryFQDN(), metadata.getPrivateIp()))
+                .findFirst().orElseThrow(() -> new TestFailException("FreeIPA upgrade DNS lookups test failed, idbroker instance group was not found."));
+        try {
+            String cmd = String.format(CHECK_DNS_LOOKUPS_CMD, instanceGroupMetadata.getDiscoveryFQDN(), instanceGroupMetadata.getPrivateIp());
+            Map<String, Pair<Integer, String>> results = sshJClientActions.executeSshCommand(getInstanceGroups(sdxTestDto, sdxClient),
+                    List.of(HostGroupType.MASTER.getName()), cmd, false);
+            results.values().forEach(result -> Assertions.assertTrue(result.getLeft() == 0));
```
Comment (lacikaaa) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/testcase/e2e/freeipa/FreeIpaUpgradeTests.java:321]: for equality you should use `assertEquals`

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/testcase/e2e/freeipa/FreeIpaUpgradeTests.java [hunk 6]
```diff
@@ -301,10 +308,27 @@ private void kinit(SdxTestDto sdxTestDto, SdxClient sdxClient) {
...
+        } catch (TestFailException e) {
```
Comment (lacikaaa) [inline integration-test/src/main/java/com/sequenceiq/it/cloudbreak/testcase/e2e/freeipa/FreeIpaUpgradeTests.java:322]: this should catch any exception, not only `TestFailException`

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/testcase/e2e/freeipa/FreeIpaUpgradeTests.java (hunk 5 above)
Comment (mhalmy) [reply integration-test/src/main/java/com/sequenceiq/it/cloudbreak/testcase/e2e/freeipa/FreeIpaUpgradeTests.java]: Done.

File: integration-test/src/main/java/com/sequenceiq/it/cloudbreak/testcase/e2e/freeipa/FreeIpaUpgradeTests.java (hunk 6 above)
Comment (mhalmy) [reply integration-test/src/main/java/com/sequenceiq/it/cloudbreak/testcase/e2e/freeipa/FreeIpaUpgradeTests.java]: Done.

**Review by lacikaaa — APPROVED** [review]

**Pull request merged**